polkadot/xcm/xcm-executor/src/lib.rs

```diff
@@ -613,79 +660,92 @@ impl<Config: config::Config> XcmExecutor<Config> {
 				Ok(())
 			},
 			DepositAsset { assets, beneficiary } => {
-				let deposited = self.holding.saturating_take(assets);
-				for asset in deposited.into_assets_iter() {
-					Config::AssetTransactor::deposit_asset(
-						&asset,
-						&beneficiary,
-						Some(&self.context),
-					)?;
+				let old_holding = self.holding.clone();
+				let result = Config::TransactionalProcessor::process(|| {
+					let deposited = self.holding.saturating_take(assets);
+					for asset in deposited.into_assets_iter() {
+						Config::AssetTransactor::deposit_asset(
+							&asset,
+							&beneficiary,
+							Some(&self.context),
+						)?;
+					}
+					Ok(())
+				});
+				if Config::TransactionalProcessor::IS_TRANSACTIONAL && result.is_err() {
+					self.holding = old_holding;
 				}
-				Ok(())
+				result
 			},
 			DepositReserveAsset { assets, dest, xcm } => {
-				// we need to do this take/put cycle to solve wildcards and get exact assets to be
-				// weighed
-				let to_weigh = self.holding.saturating_take(assets.clone());
-				self.holding.subsume_assets(to_weigh.clone());
-
-				let mut message_to_weigh =
-					vec![ReserveAssetDeposited(to_weigh.into()), ClearOrigin];
-				message_to_weigh.extend(xcm.0.clone().into_iter());
-				let (_, fee) =
-					validate_send::<Config::XcmSender>(dest.clone(), Xcm(message_to_weigh))?;
-				// set aside fee to be charged by XcmSender
-				let parked_fee = self.holding.saturating_take(fee.into());
```

Comment:
We need to make sure to keep this fix

Reply:
Oops, change got clobbered, adding it back now.